Switch to api directory.

cahrens opened the pull request:
Does the following--

1) Switch to api with additional tests
2) Improves Swagger documentation
3) Improves error handling around validation

cahrens commented:
@andy-armstrong this is ready for your review!

andy-armstrong commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -0,0 +1,219 @@
+        raise AccountUpdateError({"field_errors": field_errors})
+
+    # If everything validated, go ahead and save the serializers.
+    for serializer in user_serializer, legacy_profile_serializer:
```
Should this also be wrapped so that it raises AccountUpdateError if it fails? Currently it could raise exceptions that aren't documented.

cahrens replied:
Yes, I will wrap the remainder of the bits with an AccountUpdateError.

I wonder if I should create a specific type of error for validation errors (as opposed to passed validation, but error occurred during update). Perhaps AccountValidationError? What do you think @andy-armstrong?

andy-armstrong replied:
I think adding AccountValidationError is a great idea.

andy-armstrong commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -0,0 +1,219 @@
+            datetime.datetime.now(UTC).isoformat()
+        ])
+        existing_user_profile.set_meta(meta)
+        existing_user_profile.save()
```
This is less important, but it could also fail too I suppose.

andy-armstrong commented on openedx/core/djangoapps/user_api/accounts/tests/test_api.py:
```diff
@@ -0,0 +1,144 @@
+
+@unittest.skipUnless(settings.ROOT_URLCONF == 'lms.urls', 'Account APIs are only supported in LMS')
+class TestAccountApi(TestCase):
+    password = "test"
```
Maybe add a docstring here that mirrors the module level one. I didn't look up to the top of the file to notice that you'd described why the API is only testing errors, and I think others might make that mistake.

andy-armstrong commented:
:+1: I'm really happy with the separation out of the API. We should let other folks know what we're doing here.

kawaguchi-ks commented:
Can one of the admins verify this patch?

cahrens commented:
@andy-armstrong please review my last commit, which addresses your feedback. Note that I also simplified the form of the detailed error information in AccountUpdateError and AccountValidationError, so it is clearer what will be returned.

andy-armstrong commented:
:+1: looks great. You have a unit test failure which doesn't expect the new validation error btw.
